File: [H]array/[H]array-sliding-window/438-find-all-anagrams-in-a-string.py
from collections import defaultdict
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        pchar_freq = defaultdict(int)
        for c in p:
            pchar_freq[c] += 1
        
        res = []
        match_total = len(pchar_freq)
        match = 0
        schar_freq = defaultdict(int)
        left = 0
        for right in range(len(s)):
            schar_freq[s[right]] += 1
            if schar_freq[s[right]] == pchar_freq[s[right]]:
                match += 1
            while right - left + 1 > len(p):
                if schar_freq[s[left]] == pchar_freq[s[left]]:
                    match -= 1
                schar_freq[s[left]] -= 1
                if schar_freq[s[left]] == 0:
                    schar_freq.pop(s[left])
                left += 1
            if match == match_total:
                res.append(left)
        return res

# time O(m+n), due to traverse
# space O(1), not counting output
# using array and standard sliding window and hashmap

File: [H]array/[H]array-sliding-window/438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        HashMap<Character, Integer> pcharToFreq = new HashMap<>();
        for (int i = 0; i < p.length(); i++) {
            pcharToFreq.compute(p.charAt(i), (k, v) -> v == null ? 1 : v + 1);
        }
        HashMap<Character, Integer> subcharToFreq = new HashMap<>();
        int total = pcharToFreq.size();
        int match = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            subcharToFreq.compute(s.charAt(right), (k, v) -> v == null ? 1 : v + 1);
            if (Objects.equals(subcharToFreq.get(s.charAt(right)), pcharToFreq.get(s.charAt(right)))) {
                match += 1;
            }
            while (right - left + 1 > p.length()) {
                Character c = s.charAt(left);
                if (Objects.equals(subcharToFreq.get(c), pcharToFreq.get(c))) {
                    match -= 1;
                }
                subcharToFreq.put(c, subcharToFreq.get(c) - 1);
                if (Objects.equals(subcharToFreq.get(c), 0)) {
                    subcharToFreq.remove(c);
                }
                left += 1;
            }
            if (match == total) {
                res.add(left);
            }
        }
        return res;
    }
}

// time O(m+n), due to traverse
// space O(1), not counting output
// using array and standard sliding window and hashmap